qringbuf.go

```diff
@@ -545,8 +824,31 @@ func (qrb *QuantizedRingBuffer) Stop() (didResultInStop bool) {
 	return
 }
 
-func (qrb *QuantizedRingBuffer) Restart(readLimit int64) error {
-
+// StartFill is called at the start of every stream-cycle. If readLimit is 0
```

Comment:
See https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L833-L850
Essentially qringbuf does not care what your offset is within the reader. It just... goes forward.

This is also why I explicitly expose the mutex as stable part of the API - it allows me to "suspend everything" and jump way forward if I need to... though it now occurs to me that I need to remove the locking from Buffered() in order to make this workable.

Back to your questions above:
- Yes, otherwise NextRegion will say "dunno what to give you: a collector is not running"
- If you say 0, the only thing that changes is when the collector exits ( io.EOF ). Notably:
https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L890-L894 in combination with https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L769-L779

Suggestion how to make this... clearer?

Reply:
> This is also why I explicitly expose the mutex as stable part of the API - it allows me to "suspend everything" and jump way forward if I need to... though it now occurs to me that I need to remove the locking from Buffered() in order to make this workable.

Why do you need to _suspend_ it? That seems very strange. If you need a way to "pause", just expose a separate method and automatically resume when the user next asks for a region.

> Yes, otherwise NextRegion will say "dunno what to give you: a collector is not running"

Why not just automatically start when the user asks for a region? Personally, I'd expect the API to be:

```go
qrb.SetLimit(N) // set the limit on the number of bytes to read relative to the current offset.
```

---

Basically, forget about the internals and try to figure out what you'd like the API to look like regardless of how this library is implemented under the hood.

Why would users care about explicitly starting/stopping the collector? Why does the user even care that there is a "collector".

The user cares about:

* Closing the buffer when done with it.
* Reading slices from the buffer.
* Setting limits (not actually sure _why_?).

Reply:
THANK YOU! This is exactly the type of review/comment-thread I was looking for. Sorry for rewriting half the docs - one would need to re-read this almost from scratch.

https://github.com/ipfs/go-qringbuf/commit/9f05222fd67 resolves all remaining comments ( I marked them as such ). Notably:

> Can we put this in the description for the individual methods? The reader won't know what any of these functions do till they read the actual API.

Re-wrapped, I did a bit of both. Worry is that the reader **won't read** down to the API without an overview making them go "huh... that's... not something I've seen before..."

> I'd focus more on why and what and less on how.

Tried to do so... not sure how successful. Let's chat about this more ;)

> Avoid exposing mutexes 

This is explicitly by design - expanded upon here: https://github.com/ipfs/go-qringbuf/blob/9f05222fd67da74107b91461c9f1d3c97e160ce3/qringbuf.go#L8-L11 and here https://github.com/ipfs/go-qringbuf/blob/9f05222fd67da74107b91461c9f1d3c97e160ce3/qringbuf.go#L519-L521